[pravega] Issue 50: Allow deployment of multiple pravega clusters with external access

anishakj opened the pull request:
Signed-off-by: anishakj <anisha.kj@dell.com>

### Change log description

Currently, in the charts we are creating service account roles and role bindings if external access is enabled. But there is no check to enforce they are already present. Due to that after one cluster deployed, second cluster deployment failing with error resource already exists

### Purpose of the change

 Fixes #50

### What the code does

In charts, added a check to verify the existence of resources before  creating it

### How to verify it

Verified that able to deploy more than one cluster with external access

### Checklist

_(Place an '[x]' (no spaces) in all applicable fields)_

- [x] PR title starts with the name of the chart followed by the issue number (e.g. `[bookkeeper-operator] Issue XX: "Description"`)
- [x] Verified output of helm lint
- [x] Changes have been tested manually

nishant-yt reviewed (approved):
LGTM

SrishT reviewed (changes requested):
I feel we could add the cluster name to these RBAC components so that a different set of these is created for each cluster. In general I dont feel it would be a good practise to rely on the ones created by another chart since the deletion of both these charts would be independent of each other.

anishakj commented:
> I feel we could add the cluster name to these RBAC components so that a different set of these is created for each cluster. In general I dont feel it would be a good practise to rely on the ones created by another chart since the deletion of both these charts would be independent of each other.

In that case, if we mention the service account name in the values file, that wont be actual service account name created and there will be a mismatch because the service account name craeted is <clustername> appended with service account

SrishT commented:
> In that case, if we mention the service account name in the values file, that wont be actual service account name created and there will be a mismatch because the service account name created is appended with service account

The service account name should not be accepted from the use, instead it should be named in a way which would ensure every new deployment would use a unique name for these resources
```
{{ template "pravega.fullname" . }}-service-account
```
or something similar

anishakj commented:
> > In that case, if we mention the service account name in the values file, that wont be actual service account name created and there will be a mismatch because the service account name created is appended with service account
> 
> The service account name should not be accepted from the use, instead it should be named in a way which would ensure every new deployment would use a unique name for these resources
> 
> ```
> {{ template "pravega.fullname" . }}-service-account
> ```
> 
> or something similar

Sure, But in my testing  I have created 2 clusters. One in the deafult namespace and another one in test  namespace, but even after i deleted the first cluster, `clusterrole` didnt get delete.

anishakj commented:
> > > In that case, if we mention the service account name in the values file, that wont be actual service account name created and there will be a mismatch because the service account name created is appended with service account
> > 
> > 
> > The service account name should not be accepted from the use, instead it should be named in a way which would ensure every new deployment would use a unique name for these resources
> > ```
> > {{ template "pravega.fullname" . }}-service-account
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > or something similar
> 
> Sure, But in my testing I have created 2 clusters. One in the deafult namespace and another one in test namespace, but even after i deleted the first cluster, `clusterrole` didnt get delete.



> > > In that case, if we mention the service account name in the values file, that wont be actual service account name created and there will be a mismatch because the service account name created is appended with service account
> > 
> > 
> > The service account name should not be accepted from the use, instead it should be named in a way which would ensure every new deployment would use a unique name for these resources
> > ```
> > {{ template "pravega.fullname" . }}-service-account
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > or something similar
> 
>Will give it a try

anishakj commented:
> > > > In that case, if we mention the service account name in the values file, that wont be actual service account name created and there will be a mismatch because the service account name created is appended with service account
> > > 
> > > 
> > > The service account name should not be accepted from the use, instead it should be named in a way which would ensure every new deployment would use a unique name for these resources
> > > ```
> > > {{ template "pravega.fullname" . }}-service-account
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > or something similar
> > 
> > 
> > Sure, But in my testing I have created 2 clusters. One in the deafult namespace and another one in test namespace, but even after i deleted the first cluster, `clusterrole` didnt get delete.
> 
> > > > In that case, if we mention the service account name in the values file, that wont be actual service account name created and there will be a mismatch because the service account name created is appended with service account
> > > 
> > > 
> > > The service account name should not be accepted from the use, instead it should be named in a way which would ensure every new deployment would use a unique name for these resources
> > > ```
> > > {{ template "pravega.fullname" . }}-service-account
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > or something similar
> > 
> > 
> > Will give it a try

Changed it as suggested. Please have a look

jkhalack commented on charts/pravega/templates/clusterrole.yaml:
```diff
@@ -2,7 +2,7 @@
 metadata:
-  name: {{ .Values.serviceAccount.name }}
+  name: {{ template "pravega.fullname" . }}-{{ .Release.Namespace }}
   namespace: {{ .Release.Namespace }}
```
ClusterRole is not a namespaced resource. Any idea how the `namespace` metadata got here?

anishakj replied:
Thanks for pointing it, I can remove it

jkhalack replied:
Again, if this role is needed to deploy pravega cluster in one specific namespace, it should not be a ClusterRole at all. 
A namespaced `Role` resource should be enough here.

anishakj replied:
Roles are there in addition to clusterrole

jkhalack replied:
Sorry, did not notice it needed access to `nodes`.

jkhalack commented on charts/pravega/templates/clusterrolebinding.yaml:
```diff
@@ -2,16 +2,16 @@
 metadata:
-  name: {{ .Values.serviceAccount.name }}
+  name: {{ template "pravega.fullname" . }}-{{ .Release.Namespace }}
   namespace: {{ .Release.Namespace }}
```
Again, `ClusterRoleBinding` is a cluster-wide resource that is not confined to one namespace.

anishakj replied:
fixed

jkhalack commented on charts/pravega/templates/clusterrole.yaml:
```diff
@@ -2,8 +2,7 @@
 metadata:
-  name: {{ .Values.serviceAccount.name }}
-  namespace: {{ .Release.Namespace }}
+  name: {{ template "pravega.fullname" . }}-{{ .Release.Namespace }}
```
I would suggest having a  pravega-related  suffix here (`-pravega`?), on top of pravega cluster name and namespace.
Reasoning: There might be other components installed to the cluster that follow the pattern of `<custom-resource-name>-<namespace>`. If pravega is one of the components belonging to the same project, there is non-zero probability all custom resources would be named after the project name.

anishakj replied:
`name: {{ template "pravega.fullname" . }}-pravega-{{ .Release.Namespace }}`  like this?

jkhalack replied:
yes, like this

anishakj replied:
Made the changes

SrishT replied:
@jkhalack @anishakj I am not sure if adding `pravega` additionally to the name would be required since that is already handled by `{{ template "pravega.fullname" . }}`

jkhalack replied:
If I am not mistaken, `pravega.fullname` would contain a `pravega` substring only in the case when the user has not overriden it.

anishakj replied:
@jkhalack My understanding is if `fullnameoverride` is not used, and chart name is not pravega `pravega` is appended. But user can set the name using `fullnameoverride` in that case `pravega` is not appended. Please refer [this ](https://github.com/pravega/charts/blob/master/charts/pravega/templates/_helpers.tpl#L14-#L25)

anishakj replied:
Reverted the changes and  updated documentation

jkhalack replied:
I am fine with documenting the possibility of name clashing when using `fullnameoverride`.

SrishT reviewed (approved):
LGTM
